src/goLanguageServer.ts

```diff
@@ -0,0 +1,377 @@
+import { GoReferenceProvider } from './goReferences';
+import { GoImplementationProvider } from './goImplementations';
+import { GoTypeDefinitionProvider } from './goTypeDefinition';
+import { GoDocumentFormattingEditProvider } from './goFormat';
+import { GoRenameProvider } from './goRename';
+import { GoDocumentSymbolProvider } from './goOutline';
+import { GoSignatureHelpProvider } from './goSignature';
+import { GoWorkspaceSymbolProvider } from './goSymbol';
+import { parseLiveFile } from './goLiveErrors';
+import { promptForMissingTool } from './goInstallTools';
+import { getBinPath, getCurrentGoPath } from './util';
+
+interface LanguageServerConfig {
+	enabled: boolean;
+	flags: string[];
+	features: {
+		completion: boolean;
+		diagnostics: boolean;
+		format: boolean;
+		definition: boolean;
+		typeDefinition: boolean;
+		hover: boolean;
+		references: boolean;
+		rename: boolean;
+		signatureHelp: boolean;
+		documentSymbols: boolean;
+		workspaceSymbols: boolean;
+		implementation: boolean;
+		documentLink: boolean;
+	};
+}
+
+// registerLanguageFeatures registers providers for all the language features.
+// It looks to either the language server or the standard providers for these features.
+export function registerLanguageFeatures(ctx: vscode.ExtensionContext) {
+	// Subscribe to notifications for changes to the configuration of the language server.
+	ctx.subscriptions.push(vscode.workspace.onDidChangeConfiguration(e => watchLanguageServerConfiguration(e)));
+
+	let config = parseLanguageServerConfig();
+
+	// If the user has not enabled the language server,
+	if (!config.enabled) {
+		registerUsualProviders(ctx);
+		return;
+	}
+
+	let path = getLanguageServerToolPath();
+	const toolName = getToolFromToolPath(path);
+	const c = new LanguageClient(
+		toolName,
+		{
+			command: path,
+			args: ['-mode=stdio', ...config.flags],
+			options: {
+				env: getToolsEnvVars(),
+			},
+		},
+		{
+			initializationOptions: {},
+			documentSelector: ['go', 'go.mod', 'go.sum'],
+			uriConverters: {
+				// Apply file:/// scheme to all file paths.
+				code2Protocol: (uri: vscode.Uri): string => (uri.scheme ? uri : uri.with({ scheme: 'file' })).toString(),
+				protocol2Code: (uri: string) => vscode.Uri.parse(uri),
+			},
+			revealOutputChannelOn: RevealOutputChannelOn.Never,
+			middleware: {
+				provideDocumentFormattingEdits: (document: vscode.TextDocument, options: FormattingOptions, token: vscode.CancellationToken, next: ProvideDocumentFormattingEditsSignature) => {
+					if (!config.features.format) {
+						return [];
+					}
+					return next(document, options, token);
+				},
+				provideCompletionItem: async (document: vscode.TextDocument, position: vscode.Position, context: vscode.CompletionContext, token: vscode.CancellationToken, next: ProvideCompletionItemsSignature) => {
+					if (!config.features.completion) {
+						return [];
+					}
+					return next(document, position, context, token);
```

Comment:
When completion is coming from the language server, we still want to include some completion items from the extension like the completions for unimported packages. The code for that was present in `goMain.ts` file. Any reason for not keeping it?

Reply:
We just added support for this feature in `gopls` (just standard library completions for now, but packages in the module cache/`GOPATH` will be coming soon). I expect that this will be completed before the next release of the VSCode extension.

Reply:
Below is what comes from the extension even when language server is enabled
- Completion for user tries to type in `package`. The completion item provided is for the package name based on the folder
- Completion if you are typing the first word in a line comment right above an exported member. This simply provides the name of the exported member as the completion item
- Completion for keywords. `gocode` never had this. Does `gopls` plan to add this? It would be very straight forward to do so.
- Completion for unimported packages. When used chooses one of these items, not only is the current word completed, but an import statement gets added too

Can you share the plans on each of the above in gopls?

I would prefer we keep the above for now, and then remove them one by one as `gopls` adds support for them

Reply:
Completion for comments, keywords, and package names we could probably do fairly simply in `gopls`. I will file feature requests for these. Unimported packages is coming in the next version of `gopls`.

My reasoning for removing this code is primarily because `gopls` has a fair bit of logic for ranking completions, and adding additional completions from the extension will likely mess with the ordering in some way. Also, having this combination of completions hides the areas in which `gopls` is lacking - I wouldn't have known about these additional features if you hadn't just explained them to me right now. If we want to move all completion logic into `gopls`, I think we need begin the process of noticing where `gopls` is lacking. Maybe as a compromise we could add a temporary setting to force completions to be `gopls`-only?

Reply:
My only concern with this approach is the noise of incoming GitHub issues saying that the completion for unimported packages is missing when not using modules. But if you believe this feature will get added to gopls before we release an update to this extension, then we should be good as the answer to such issues would be to update gopls